File: src/main/java/io/github/takenoha/towerdefense/paper/CoreMaterialPolicy.java
package io.github.takenoha.towerdefense.paper;

import java.util.Objects;
import org.bukkit.Material;

/**
 * The physical materials used by the durable core identity.
 *
 * <p>Core identity is still PDC-backed for inventory items, but the placed block is identified
 * by the database-backed core coordinate.  This is necessary because the current placed block,
 * dried kelp, is not a tile entity and cannot carry a persistent data container.</p>
 */
public final class CoreMaterialPolicy {
    public static final Material CURRENT_ITEM = Material.RESIN_BRICKS;
    public static final Material LEGACY_ITEM = Material.NETHER_STAR;
    public static final Material CURRENT_BLOCK = Material.DRIED_KELP_BLOCK;
    public static final Material LEGACY_BLOCK = Material.BEACON;

    private CoreMaterialPolicy() {
    }

    public static boolean isCoreItemMaterial(Material material) {
        return material == CURRENT_ITEM || material == LEGACY_ITEM;
    }

    public static boolean isLegacyItemMaterial(Material material) {
        return material == LEGACY_ITEM;
    }

    public static boolean isCoreBlockMaterial(Material material) {
        return material == CURRENT_BLOCK || material == LEGACY_BLOCK;
    }

    public static boolean isCurrentBlock(Material material) {
        return material == CURRENT_BLOCK;
    }

    public static void requireCoreItemMaterial(Material material) {
        if (!isCoreItemMaterial(Objects.requireNonNull(material, "material"))) {
            throw new IllegalArgumentException("unsupported core item material: " + material);
        }
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/paper/CoreMaterialPolicy.kt
package io.github.takenoha.towerdefense.paper

import java.util.Objects
import org.bukkit.Material

/** Physical materials used by the durable core identity. */
class CoreMaterialPolicy private constructor() {
    companion object {
        @JvmField
        val CURRENT_ITEM: Material = Material.RESIN_BRICKS

        @JvmField
        val LEGACY_ITEM: Material = Material.NETHER_STAR

        @JvmField
        val CURRENT_BLOCK: Material = Material.DRIED_KELP_BLOCK

        @JvmField
        val LEGACY_BLOCK: Material = Material.BEACON

        @JvmStatic
        fun isCoreItemMaterial(material: Material?): Boolean =
            material == CURRENT_ITEM || material == LEGACY_ITEM

        @JvmStatic
        fun isLegacyItemMaterial(material: Material?): Boolean = material == LEGACY_ITEM

        @JvmStatic
        fun isCoreBlockMaterial(material: Material?): Boolean =
            material == CURRENT_BLOCK || material == LEGACY_BLOCK

        @JvmStatic
        fun isCurrentBlock(material: Material?): Boolean = material == CURRENT_BLOCK

        @JvmStatic
        fun requireCoreItemMaterial(material: Material?) {
            val required = Objects.requireNonNull(material, "material")
            if (!isCoreItemMaterial(required)) {
                throw IllegalArgumentException("unsupported core item material: $required")
            }
        }
    }
}
